Add CloudFoundry Resource Attributes

KarstenSchnitter opened the pull request:
Proposal for #622:

Using the CloudFoundry documentation, provides a set of resource attribute describing CloudFoundry workloads. Both applications as well as system components are addressed. Helpful resources:

* https://github.com/cloudfoundry/loggregator-api#v2-envelope
* https://docs.cloudfoundry.org/devguide/deploy-apps/environment-variable.html#VCAP-APPLICATION
* https://bosh.io/docs/jobs/#properties-spec
* https://docs.cloudfoundry.org/devguide/deploy-apps/streaming-logs.html

## Changes

Introduces a resource attributes convention for CloudFoundry. Defines resource attributes to identify CloudFoundry applications and system components, enabling and improving OpenTelemetry support. The proposed attributes are selected to provide an as small as possible payload. The changes follow the proposal in #622.

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [x] [CHANGELOG.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CHANGELOG.md) updated for non-trivial changes.
* [x] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.

joaopgrassi commented:
@KarstenSchnitter 

> A changelog entry seems to be necessary

Yes, please add an entry to the CHANGELOG.md file. You can see an example here: https://github.com/open-telemetry/semantic-conventions/pull/615

> I do not know, how to verify the contribution guidelines. Locally a make check ran without issues.

I approved the CI run and some checks are failing. Not sure how the `make check` passed, but you can run them separately and they for sure should also fail locally.

KarstenSchnitter commented:
@joaopgrassi I will take care of the issues and the changelog. Is there anything else?

KarstenSchnitter commented:
@joaopgrassi I noticed some unmerged changes on my side. Hopefully, this resolves the checks.

joaopgrassi commented:
I only looked at a high level yesterday, and I'm not experienced with CloudFoundry, so I will need to read a bit to understand the context so I can review it. I will try to do it in the upcoming days.

FWinkler79 commented on docs/attributes-registry/cloudfoundry.md:
```diff
@@ -0,0 +1,90 @@
+| `cloudfoundry.org.name` | string | The name of the CloudFoundry organization the app is running in. [5] | `my-org-name` |
+| `cloudfoundry.process.id` | string | The UID identifying the process. [6] | `218fc5a9-a5f1-4b54-aa05-46717d0ab26d` |
+| `cloudfoundry.process.type` | string | The type of process. [7] | `web` |
+| `cloudfoundry.source_id` | string | A guid or another name describing the event source. [8] | `218fc5a9-a5f1-4b54-aa05-46717d0ab26d`; `cf/gorouter` |
```
If there is a `source_id` and a `source_type`, shouldn't the naming rather be: `source.id` and `source.type`, i.e. `source` being a grouping prefix?

FWinkler79 commented on docs/attributes-registry/cloudfoundry.md:
```diff
@@ -0,0 +1,90 @@
+|---|---|---|---|
+| `cloudfoundry.app.id` | string | The guid of the application. [1] | `218fc5a9-a5f1-4b54-aa05-46717d0ab26d` |
+| `cloudfoundry.app.name` | string | The name of the application. [2] | `my-app-name` |
+| `cloudfoundry.instance_id` | string | An index or guid describing the concrete instance of the event source. [3] | `0`; `218fc5a9-a5f1-4b54-aa05-46717d0ab26d` |
```
Does an instance only have an `id` or are there other instance-related properties?
If the latter is the case, I would suggest a different naming: `cloudfoundry.instance.id`

KarstenSchnitter replied:
This was resolved by splitting the instance id into `app.instance.id` and `system.instance.id`.

FWinkler79 commented on docs/resource/cloudfoundry.md:
```diff
@@ -0,0 +1,167 @@
+<!-- semconv cloudfoundry(tag=generic) -->
+| Attribute  | Type | Description  | Examples  | Requirement Level |
+|---|---|---|---|---|
+| [`cloudfoundry.instance_id`](../attributes-registry/cloudfoundry.md) | string | An index or guid describing the concrete instance of the event source. [1] | `0`; `218fc5a9-a5f1-4b54-aa05-46717d0ab26d` | Required |
```
see comment above.

FWinkler79 commented on docs/resource/cloudfoundry.md:
```diff
@@ -0,0 +1,167 @@
+| Attribute  | Type | Description  | Examples  | Requirement Level |
+|---|---|---|---|---|
+| [`cloudfoundry.instance_id`](../attributes-registry/cloudfoundry.md) | string | An index or guid describing the concrete instance of the event source. [1] | `0`; `218fc5a9-a5f1-4b54-aa05-46717d0ab26d` | Required |
+| [`cloudfoundry.source_id`](../attributes-registry/cloudfoundry.md) | string | A guid or another name describing the event source. [2] | `218fc5a9-a5f1-4b54-aa05-46717d0ab26d`; `cf/gorouter` | Required |
```
see comment above.

FWinkler79 commented on docs/resource/cloudfoundry.md:
```diff
@@ -0,0 +1,167 @@
+<!-- semconv cloudfoundry(tag=source_type) -->
+| Attribute  | Type | Description  | Examples  | Requirement Level |
+|---|---|---|---|---|
+| [`cloudfoundry.source_type`](../attributes-registry/cloudfoundry.md) | string | The type of the event source. [1] | `APP/PROC/WEB/0`; `APP/TASK/jobscheduler-task-70013817-ab92-4f75-82d4-de8ae4dd64b4`; `RTR`; `STG` | Recommended |
```
see comment above.

FWinkler79 reviewed (changes requested):
I think we should review the naming of some of the attributes.
The OTEL colleagues should weigh in here, as they probably know best about the usual naming conventions.

chombium commented:
Hi @FWinkler79, 
thanks for the tips and the review. I'm one of the maintainers in the Cloud Foundry's logging and metrics team. We've discussed with @KarstenSchnitter  and @ctlong and we also had the same doubts about the `source_id`, `source_type` and `instance_type` labels. We've added them as they are currently used in the Cloud Foundry's logging and metrics stack, but after checking the existing conventions, I've found the [attributes-registry/source.md](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/source.md) and [general/attributes.md](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/attributes.md) and I think that we should changed them to `source.id`, `source.type` and `instance.id`.

There is one more open question to clear: In Cloud Foundry the `source type` label is only used for logs and not for metrics and traces. We are not sure what do do about it and if we need to define it in some specific way. @joaopgrassi Any ideas about this?

KarstenSchnitter commented:
I provided a revised version with the fields renamed as suggested by @FWinkler79. Furthermore, I added some clarification on the `source_type`, especially the requirement level. For the three use-cases in #622 I added a comment, how to obtain the appropriate value for use-case 1, where the value needs to be set to a static value. This is the reason, why I think `source.type`should be part of this convention, but only apply to logs as now reflected in the requirement level. The branch to be merged was rebased on main in the process.

FWinkler79 commented:
Thanks for your reply and changes.

@chombium: 
> There is one more open question to clear: In Cloud Foundry the source type label is only used for logs and not for metrics and traces. We are not sure what do do about it and if we need to define it in some specific way. @joaopgrassi Any ideas about this? 

I am not 100% sure, but I recently updated our internal version of the OTEL SemConv to the latest version and noticed that the attributes for different signal types are now (rather) cleanly separated into log-attributes, trace-attributes, etc.
Not sure the documentation already properly reflects that, but at least the YAML specifications do.
E.g. you can find the entire yaml-based model files [here](https://github.com/open-telemetry/semantic-conventions/tree/main/model). There you find folders for logs, metrics, traces and recource attributes.
To avoid data duplication the common "base attributes" that are shared between signal types are now maintained under the "registry" folder. It's not entirely cleaned up yet, but I think that's the intention.

For example, in the `registry` folder you can find `http.yaml`, in the `trace` folder you also find [`http.yaml`](https://github.com/open-telemetry/semantic-conventions/blob/main/model/trace/http.yaml). The trace-related HTTP attributes simply `ref`erence HTTP attributes defined in the file under `registry`.

In a similar way, I think you could define log-specific attributes.

chombium commented:
Thanks for the prompt change @KarstenSchnitter and for the suggestions and digging into the definitions @FWinkler79. @joaopgrassi  Is there something else to be done apart from the suggestions about the `source type` parameter?

joaopgrassi commented on model/registry/cloudfoundry.yaml:
```diff
@@ -0,0 +1,140 @@
+groups:
+  - id: registry.cloudfoundry
+    prefix: cloudfoundry
+    type: resource
```
If we have a mix of resource/signal attributes in the registry, I think we need to mark these as an attribute group.

```suggestion
    type: attribute_group
```

We have an issue open that is tracking changes we need to make in the tooling to support a mix like this https://github.com/open-telemetry/build-tools/issues/242. A PR was merged recently but I'm not sure if that alone solves everything. Please @lmolkova @AlexanderWert could you confirm if that's the case?

KarstenSchnitter replied:
Currently, all attributes are proposed to be resource attributes. I keep this suggestion in mind for the discussion about the `source.type`.

KarstenSchnitter replied:
I made that change in the last commit after merging the current main branch into this PR.

joaopgrassi commented on model/registry/cloudfoundry.yaml:
```diff
@@ -0,0 +1,140 @@
+        examples:
+          - '218fc5a9-a5f1-4b54-aa05-46717d0ab26d'
+          - 'cf/gorouter'
+      - id: instance.id
```
> An index or guid describing the concrete instance of the event source.

Is the `instance.id` referring to the actual app/resource? Or referring to something else? This part is a bit confusing, but maybe it's just my lack of CloudFoundry knowledge

> It is supposed to contain the application instance index for applications deployed on the runtime or the vm id for CloudFoundry components.

ctlong replied:
`instance.id` refers to the VM that the app or system component is running on.

KarstenSchnitter replied:
This already is confusing in CloudFoundry: The instance.id can either be the app instance index starting from 0, if an app is deployed with multiple instances (similar to the index of a K8s StatefulSet). Or it can be the guid of a vm running a system component.

Picture the log stream of a typical CF application. It will contain logs of the application itself, where the instance id just contains the app instance index. But there will also be access logs from the Gorouter (ingress load balancer) which will contain the vm guid in the instance id. Consumers of CloudFoundry logs and metrics are used to this duality of use.

joaopgrassi commented on model/registry/cloudfoundry.yaml:
```diff
@@ -0,0 +1,140 @@
+          variable `VCAP_APPLICATION.application_name`. This is the same value
+          as reported by `cf apps`.
+        examples: ['my-app-name']
+      - id: app.id
```
Please have a look at this PR https://github.com/open-telemetry/semantic-conventions/pull/312. It seems this is exactly the same there.

ctlong replied:
`app.id` is different than `instance.id`. The former is the GUID of the app that the signal originates from, and the latter refers to the VM that the app or system component is running on.

KarstenSchnitter replied:
@ctlong to my understanding of https://github.com/cloudfoundry/loggregator-api#v2-envelope, the instance id will be the app instance index and not a cell or vm id for applications. The GUID is only used for system components such as the Gorouter. This is also my experience with these kind of signals.

joaopgrassi commented on model/registry/cloudfoundry.yaml:
```diff
@@ -0,0 +1,140 @@
+          variable `VCAP_APPLICATION.org_id`. This is the same value as
+          reported by `cf org <org-name> --guid`.
+        examples: ['218fc5a9-a5f1-4b54-aa05-46717d0ab26d']
+      - id: source.type
```
I'm a bit confused by the `source.type` attribute. Is it changing over the course of the application? Is it a fixed value known when starting up? All the attributes here are defined as resources, which mean, they should be readily available when the application/SDK is initializing. I saw in the discussion that you were not sure if this is supposed to be a log/metric vs resource attribute?

KarstenSchnitter replied:
For any running component, the source.type will be fixed. However, a CloudFoundry applications can consist of multiple processes, that will have different source.type fixed at their respective deployment type. CloudFoundry will report all those components under the same app.id. Picture a K8s pod consisting of several containers. 

In addition CloudFoundry will add logs from system components with different source.types. Examples are Gorouter access logs with source.type=RTR or staging logs with source.type=STG. See https://docs.cloudfoundry.org/devguide/deploy-apps/streaming-logs.html#format for more examples. In each case, the source.type remains constant during runtime for the emitting component. Note, that even if those components emit logs for different applications, e.g. the Gorouter, the source.type remains fixed for the particular component.

ctlong replied:
`source.type` is used to differentiate different log signals. It may be populated by both applications and system components. Here's the Cloud Foundry docs about it: https://docs.cloudfoundry.org/devguide/deploy-apps/streaming-logs.html#format.

Just considering applications, the `source.type` attribute does change over the course of the application. When an application is being staged, the `source.type` is set to `STG/0`. Once an application has started up the `source.type` is set to `APP/...`.

In Cloud Foundry, "applications" are comprised of many "processes" (e.g. "web", "worker", etc,), which each have some number of instances. The `source.type` will be populated differently depending on the application/process/instance that the signal originates from.

There's also tasks, but maybe we can ignore that for now.

ctlong replied:
(My bad @KarstenSchnitter I was typing at the same time and didn't see your message before I sent mine 😅)

joaopgrassi replied:
So if source type is changing, we have a problem here as resource attributes are suppose to be immutable, meaning a resource is created and populated on app startup and after that it does not change.

KarstenSchnitter replied:
The source type is fixed, once an application is started. It is just a closer description of the kind of resource. The source.type `STG/0` is not emitted by the application, but by the CloudFoundry runtime. Every event emitted by the application will be of the same fixed `source.type`. If there are multiple processes running, each will have its own `process.id` and possibly a custom `source.type`. However, once a process is started, the `source.type` is fixed just like the `process.id`.

joaopgrassi commented on model/registry/cloudfoundry.yaml:
```diff
@@ -0,0 +1,140 @@
+          - 'APP/TASK/jobscheduler-task-70013817-ab92-4f75-82d4-de8ae4dd64b4'
+          - 'RTR'
+          - 'STG'
+      - id: process.id
+        type: string
+        brief: >
+          The UID identifying the process.
+        note: |
+          Application instrumentation should use the value from environment
+          variable `VCAP_APPLICATION.process_id`. It is supposed to be equal to
+          `VCAP_APPLICATION.app_id` for applications deployed to the runtime.
+          For system components, this could be the actual PID.
+        examples: ['218fc5a9-a5f1-4b54-aa05-46717d0ab26d']
+      - id: process.type
+        type: string
+        brief: >
+          The type of process.
+        note: |
+          CloudFoundry applications can consist of multiple jobs. Usually the
+          main process will be of type `web`. There can be additional background
+          tasks or side-cars with different process types.
+        examples: ['web']
```
I wonder if `process.id` and `process.type` could be general process attributes https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/process.md#process

KarstenSchnitter replied:
The CloudFoundry process types have very particular values. The do not indicate the executable nor runtime. I do not find a matching field in the process resource attributes.

chombium replied:
Hi @joaopgrassi, thank you very much for the tips and the review. Tbh, I've always hated the different meanings and different values of some tags depending on the context be it telemetry data type: gauge, counter, log etc or the source CF apps or platform components. It's a technical debt that we have in Cloud Foundry. We cannot fix it for our current consumers and current formats, but we can clean up the things the consumption with OpenTelemetry :)

IMHO, we should define everything properly according to the other existing rules and semantic conventions in OpenTelemetry, so that we remove the dual meaning and the confusion with some tags and add a custom reader or processor in the Otel Collector which we deliver with CF to read the original CF formatted data and reformat everything properly according to the semantic conventions which we will define with this PR. 

I suggest that we (me, @ctlong and the other maintainers in the Logging and Metrics Team in Cloud Foundry) sit together, take a look what we have and decide how to handle the different parameters. We will than come back with all the meanings of different values for some parameters and possibly propose proper Otel attributes or we'll ask for help.

@ctlong I hope you agree with my suggestion, I've already started collecting and analyzing telemetry data from the platform components and some apps. I'll contact you when I have gathered all the needed data.

KarstenSchnitter replied:
@chombium , @ctlong: If you discuss, will you provide a suggestion about `source.id` and `instance.id` as well? After all, these two attributes also carry a dual meaning.

KarstenSchnitter replied:
Since @chombium approved this PR, I resolve this issue for now.

KarstenSchnitter commented:
@chombium , @ctlong, can you provide a status of your deliberations? Maybe we can agree on an initial convention and improve over time?

joaopgrassi commented:
Hi @KarstenSchnitter !

We changed how the CHANGELOG.md is managed. Please take a look at https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#adding-a-changelog-entry to see what needs to be done. Sorry for the disruption.

KarstenSchnitter commented:
> Hi @KarstenSchnitter !
> 
> We changed how the CHANGELOG.md is managed. Please take a look at https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#adding-a-changelog-entry to see what needs to be done. Sorry for the disruption.

To do that, i needed to merge the current main branch into the PR. I pushed the relevant changes for this.

chombium commented:
Hi, I'm still analyzing and organizing tags, so that we have good starting point with the convention

joaopgrassi commented:
Thanks for the update @chombium. Let's see if the stale bot removes the label as it should. We had some issues with it in the past so I will keep it as is to see.

chombium commented:
Hi,

In the past few weeks I was analyzing tag/field names and values from logs and metrics from system components and applications in Cloud Foundry and a part from the `source_id` and `instance_id` I think everything is clear. The problem with these two parameters is that for:
- CF system components:
     - `source_id` is a string, process/component. E.g.: `"source_id": "syslog_agent"`
     - `instance_id` is a GUID of the VM where the process is working. 
                    E.g.: `"instance_id": "77d46af7-24d3-42b7-87d9-6365cf0af281"`
- CF applications:
     - `source_id` is the GUID of the application. E.g.: `"source_id": "50cd77c8-f772-4223-9637-bdd9bee8bdee"`
     - `instance_id` is the index of the application instance. E.g.: `"instance_id": "3"`

We should think how do we represent the different origin of these values:
1. If we use the same parameters for both data from applications and system components, we'll have to think how do we process the telemetry data if we need to check if it comes from apps or system components. Maybe a custom Otel receiver or processor.
2. Use separate fields for the application and system component data, so that the data origin is clearly described without extra processing and value inspection.

I'm personally in favor of the second approach. We can do the value inspection in the existing Cloud Foundry components, so that when the data is received in the Otel part everything is clear. As we already convert and prepare the data already for ingestion by the Otel collector I guess that won't be a problem. Though, I'm not sure how should we depict this division in the semantic conventions...

@ctlong What do you think about this?

ctlong commented:
I agree that the second approach you described (separate fields) sounds more in line with what I'd expect from OTLP semantic conventions, and it makes processing easier.

I'm concerned that splitting the fields up will make translating from/to the existing loggregator format more difficult. However, maybe that shouldn't be the focus? Or maybe there's some ways with schemas, etc. that we could mitigate that problem.

chombium commented:
@ctlong I've started a discussion on Slack in [#otel-semantic-conventions-wg](https://cloud-native.slack.com/archives/C041APFBYQP/p1711402749937679) and it seems that there is some application specific schema in the works. 

From what I've understood so far, when the OTEP is finished we could have some mapping like:
- for cf apps: `source_id_from_app_metadata => cloudfoundry.app.source_id`
- for cf platform/system components: `source_id_from_system_component => cloudfoundry.system.source_id`

@joaopgrassi, @lmolkova, @AlexanderWert: Now that we have some rough idea and conclusion from the Cloud Foundry side where do we want to go, could you be kind and help us on how should we depict this in the semantic conventions?

KarstenSchnitter commented:
@chombium: How do you want to continue with the `source_id` and the `instance_id`? I like separating them by `cloudfoundry.app` and `cloudfoundry.system`. If you think, that is feasible, I can change the PR to reflect that. In my experience, the application log stream can separate the cases by the `source_type`. Everything `APP*` would be `cloudfoundry.app.source_id`, everything else would be `cloudfoundry.system.source_id`, e.g. `RTR`. Or am I missing something here.

I do not think, that `cloudfoundry.app.source_id` is actually needed, since it would be the same as the already existing `cloudfoundry.app.id`.

chombium commented:
Hi @KarstenSchnitter,

> How do you want to continue with the source_id and the instance_id? I like separating them by cloudfoundry.app and cloudfoundry.system

Let's create separate fields for both:
- source_id:
   - app => cloudfoundry.app.id
   - system component => cloudfoundry.system.id
- instance_id:
   - app => cloudfoundry.app.instance_id
   - system component => cloudfoundry.system.instance_id
   
   We, on the CF side will have to take care about filling the proper fields as the duality is on our side.
   
   @ctlong Any comments on this?
   
   >  In my experience, the application log stream can separate the cases by the source_type. Everything APP* would be cloudfoundry.app.source_id, everything else would be cloudfoundry.system.source_id, e.g. RTR. Or am I missing something here.

The APP logs origin from the app itself and the others like RTR, STG etc. origin from CF system components, but they are related to the app and that's why they come in the app's log stream. We could think about separating things later, for now we can concentrate on the format and the semantic conventions. ;)

ctlong commented:
> * instance_id:
>   * app => cloudfoundry.app.instance_id
>   * system component => cloudfoundry.system.instance_id

@chombium do we want to keep consistency with the rest of the attributes by using period separators instead of underscores? e.g. `cloudfoundry.app.instance.id` and `cloudfoundry.system.instance.id`?

chombium commented:
> do we want to keep consistency with the rest of the attributes by using period separators instead of underscores? e.g.

@ctlong Good point. I guess going forward with period as separator is better as we might add some extra instance related meta data.

KarstenSchnitter commented:
I will prepare a commit reflecting the changes proposed by @ctlong. I like the dot separation better myself.

chombium commented:
We are still working on the PR and @KarstenSchnitter will adjust the changes as discussed.

chombium reviewed (approved):
LGTM! Everything is done properly now from Cloud Foundry side. Thanks for the great work an the cooperation @KarstenSchnitter ;)

lmolkova commented on docs/attributes-registry/README.md:
```diff
@@ -38,6 +38,7 @@ Currently, the following namespaces exist:
 - [Client](client.md)
 - [Cloud](cloud.md)
 - [CloudEvents](cloudevents.md)
+- [Cloudfoundry](cloudfoundry.md)
```
nit: 

```suggestion
- [CloudFoundry](cloudfoundry.md)
```

lmolkova commented on model/registry/cloudfoundry.yaml:
```diff
@@ -0,0 +1,158 @@
+          `spec.deployment/spec.name`.
+        examples:
+          - 'cf/gorouter'
+      - id: system.instance.id
```
any reason  not to reuse existing attribute `service.instance.id` ? see https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/service.md

KarstenSchnitter replied:
One more thing: Eventually, you would want to have an attribute to identify the CloudFoundry deployment as a whole, to distinguish between different installations. Maybe that would be a proper use for `service.namespace`,

lmolkova commented on model/registry/cloudfoundry.yaml:
```diff
@@ -0,0 +1,158 @@
+          should be used. The `system.instance.id` should be set to `spec.id`.
+        examples:
+          - '218fc5a9-a5f1-4b54-aa05-46717d0ab26d'
+      - id: app.name
```
is it similar to the `service.name` ?

KarstenSchnitter replied:
Just to summarise the arguments of the ongoing conversation: For CF applications, it is very likely, that this `app.name` should also be used as the `service.name`. There might be deviations from that, e.g., including app instance index or org and space names. Furthermore, the `service.name` might be completely different for other reasons. Therefore, I suggest to provide a recommendation to use the `app.name` for the `service.name` but allow for a different value. To save data volume, the `app.name` can be omitted. This is actually true for all `*.name` attributes, if their `*.id` counter parts are there and vice versa. 

@lmolkova would you be alright with keeping this field as optional and adding the recommendation for the service name?

lmolkova commented on model/registry/cloudfoundry.yaml:
```diff
@@ -0,0 +1,158 @@
+          variable `VCAP_APPLICATION.application_id`. This is the same value as
+          reported by `cf app <app-name> --guid`.
+        examples: ['218fc5a9-a5f1-4b54-aa05-46717d0ab26d']
+      - id: app.instance.id
```
maybe this one is similar to `service.instance.id` ?

KarstenSchnitter replied:
CloudFoundry does not use unique guid or each app instance. Instead, a single guid is created for the entire instance group known as `app.id`. If you run multiple instances of the app, they will have different instance indices as explained below. This index is captured in `app.instance.id`.

This entire approach differs from OpenTelemetry's [service attributes](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/service.md). Therefore, different fields were used.

Maybe it would be clearer if `app.instance.id` was renamed to `app.instance.index`? What do you think?

lmolkova commented on model/registry/cloudfoundry.yaml:
```diff
@@ -0,0 +1,158 @@
+        examples:
+          - '0'
+          - '1'
+      - id: space.name
```
is it similar to common `service.namespace` attribute ?

KarstenSchnitter replied:
This is really something different. See my explanation in the app name. The OpenTelemetry service attributes really do not fit well with the CloudFoundry trio of org -> space -> app to describe application workloads.

lmolkova commented on model/registry/cloudfoundry.yaml:
```diff
@@ -0,0 +1,158 @@
+          - 'APP/TASK/jobscheduler-task-70013817-ab92-4f75-82d4-de8ae4dd64b4'
+          - 'RTR'
+          - 'STG'
+      - id: process.id
```
is it similar to the generic [`process.pid`](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/process.md) attribute?

KarstenSchnitter replied:
This is really not a pid. It is a guid given by the CloudFoundry runtime. A CloudFoundry application can consist of multiple [processes](https://v3-apidocs.cloudfoundry.org/version/release-candidate/#processes), similar to a K8s pod, that can contain multiple containers. This guid marks the concrete process emitting the signals. Earlier, I mentioned the trio org, space, app. this is sufficient, when there is only one process in the app. If there are more, the trio extends to a quartet: org -> space -> app -> process

lmolkova commented on model/registry/cloudfoundry.yaml:
```diff
@@ -0,0 +1,158 @@
+          variable `VCAP_APPLICATION.space_id`. This is the same value as
+          reported by `cf space <space-name> --guid`.
+        examples: ['218fc5a9-a5f1-4b54-aa05-46717d0ab26d']
+      - id: org.name
```
or maybe this one is similar to `service.namespace` ?

KarstenSchnitter replied:
See comment above.

lmolkova reviewed (commented):
Thank you for the contribution! 
I left some comments - please leverage existing attributes when they are applicable.

KarstenSchnitter reviewed (commented):
@lmolkova thanks for your review. I have tried to answer your comments as best as I could. Applications deployed to CloudFoundry are described by this hierarchy:

org (id, name) -> space (id, name) -> app (id, name, index) -> process (type, id)

I tried to outline, why this should get its own resource attribute set as proposed with this PR.

FWinkler79 reviewed (approved):
LGTM. I think these changes make sense and are helpful.

One more comment on the differentiation from `service.name`, `service.namespace` and `service.instance.id`: 

In CF, services are really very much a logical construct. CF apps are the implementations of services, and multiple integrated CF apps might constitute a single CF service. 
Also, a CF app is the underlying piece of code that serves potentially multiple service instances. A service instance is then usually a logical "shard" of one or more CF apps  reserved for and utilized by one or more service (instance) consumers. 
With that model in mind, it makes a lot of sense to separate CF apps and their related attributes from OTEL's `service.*` attributes. The former should be used to describe the pieces of software that provide the service, the latter could be used to describe the logical service construct and the instances that exist of it.

This may be a bit abstract, but hopefully it is understandable.

ctlong reviewed (approved):
LGTM. I think it will be great to have a formal declaration of CF attributes as a starting point 😁

ctlong commented:
I was recently made aware of how important `service.name` is to many of the OTel libraries – when using the OTel instrumentation, this attribute _must_ be provided or it will be set to "unknown" or something. Given that, and although I agree with @FWinkler79, I think it would be useful to later add a recommendation for `service.name` at the app/system component level of CF.

oppegard commented:
Regarding the importance of `service.name` and it being the one [required attribute](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/README.md#service) in OTel semantic conventions: I agree that it would be good to have a strong opinion on this in two CF contexts of a system component (process on BOSH VM) vs a user app. If `service.name` isn't set and a component or app embeds an OTel SDK, the default value will be "unknown_service".

Based on my understanding of the semantic meaning of `service.name` in OTel (that of a single microservice in a distributed system), my suggestion would be:

- In a system component context, `service.name` for a process should be set to the value shown in `bosh instances --ps`. For example, it would be set to "[cloud_controller_ng](https://github.com/cloudfoundry/capi-release/blob/36822440446680f4d424ab49e15d188897744d57/jobs/cloud_controller_ng/monit#L1)", "[nginx_cc](https://github.com/cloudfoundry/capi-release/blob/36822440446680f4d424ab49e15d188897744d57/jobs/cloud_controller_ng/monit#L29)", "[gorouter](https://github.com/cloudfoundry/routing-release/blob/d85f09c506f642eb64cb6af31d6d2d39426dbc02/jobs/gorouter/monit#L1)", "[tcp_router](https://github.com/cloudfoundry/routing-release/blob/d85f09c506f642eb64cb6af31d6d2d39426dbc02/jobs/tcp_router/monit#L1)", etc.
- In a user application context, `service.name` should be set to [$VCAP_APPLICATION.application_name](https://docs.cloudfoundry.org/devguide/deploy-apps/environment-variable.html#VCAP-APPLICATION), e.g. "dora". I do wonder if the app's process type should be appended, to differentiate telemetry coming from a web process vs worker. E.g. "dora:web" and "dora:worker".

chombium commented:
@oppegard You've summed up the name duality in the CF telemetry data perfectly :)
That's why we suggest that by convention the `service.*` attributes are used for CF's platform components and add another `cloudfoundry.application.*` attribute set for the applications running in Cloud Foundry. This way we can clearly separate what does the telemetry data refer to. Fwiw. I've been working with Loggretor (Cloud Foundry's Logging and Metrics Stack) for quite a while, and I've always found the dual meaning of some fields really confusing, especially with the `source_id` and `instance_id` fields. One has to always think what kind of metrics or logs they are dealing with in order to make some sense of the data :-/ 

>  I do wonder if the app's process type should be appended, to differentiate telemetry coming from a web process vs worker. E.g. "dora:web" and "dora:worker".
This could be adopted easily as a convention when we fill the fields. I haven't seen anything other than web process, though

FWinkler79 commented:
@chombium,
> That's why we suggest that by convention the service.* attributes are used for CF's platform components and add another cloudfoundry.application.* attribute set for the applications running in Cloud Foundry.

I am not sure, I would agree here. Do you suggest using `service.*` exclusively for CF platform components, and something else (i.e. the `cloudfoundry.application.*`attributes) for CF apps? 

Personally, I would be more with @oppegard on using `service.*` attributes for both platform and application components. However, I would suggest a different usage in the case of CF application components:

As mentioned before, a CF service could consist of N CF applications that are integrated to work together and are exposed as a logical (micro-)service using a service broker. As such, there is no 1:1 relationship between a CF app and a service.
So, what I would suggest is to use the `service.*` attributes to really describe the (micro-) service while the `cf.application.*` attributes describe the individual CF apps that contribute to the service.
With that given, multiple CF apps might send the same `service.name`, but different `cf.application.*` attribute values.
The `service.name` could then even be used as a correlator for logs, metrics, traces belonging to the same logical service.
A `service.instance.id` would then refer to a service _instance_, which usually is related to a specific service consumer / tenant.

What would you think about that?

chombium commented:
Hi @FWinkler79,


> I am not sure, I would agree here. Do you suggest using service.* exclusively for CF platform components, and something else (i.e. the cloudfoundry.application.*attributes) for CF apps?

Sorry for the missunderstanding caused, we are discussing this for really long time and sometimes I get lost :-/
What you describe is reasonable and I completely agree with it if we talk about applications running in CF which can be grouped in bigger solutions (I call them solutions, you've written services, but I want to avoid the confusion as in CF there is a concept of [services](https://docs.cloudfoundry.org/services/overview.html), which are basically some resources running outside CF which can be used by a CF application).

> Personally, I would be more with @oppegard on using service.* attributes for both platform and application components. However, I would suggest a different usage in the case of CF application components:

I'm not sure about using the same `service.*` attribute group for different entities CF apps and CF platform(system) components. We are proposing `cloudfoundry.system.*` attribute group for platform components and `cloudfoundry.app.*` for 
CF applications.

I understand a service as it is described in Otel as a group of applications deployed somewhere running together to fulfill some goal. I see the runtime environment where the apps are running as a separate entity.

I think that it would be nice to have separate attributes for the platform components and the apps. How are the things divided platform vs app load in Kubernetes? I see that there is a separate [k8s](https://opentelemetry.io/docs/specs/semconv/resource/k8s/) resource... Wdyt, should we mix platform components and app data in the same attribute groups?


> With that given, multiple CF apps might send the same service.name, but different cf.application.* attribute values.
The service.name could then even be used as a correlator for logs, metrics, traces belonging to the same logical service.
A service.instance.id would then refer to a service instance, which usually is related to a specific service consumer / tenant.
> What would you think about that?

In Cloud Foundry itself there is no such concept as a service in the way you are describing it. In Cloud Foundry there are only apps which are scoped in spaces and one or many spaces form an organizations. The people deploying apps in CF can have their own app grouping in services or whatever they like, but CF as platform doesn't know anything about that. All these things will be responsibility of the apps to set properly some attributes or even custom headers. All these things are nice to have, but they are not the responsibility of CF. We could write some best practice guide in the future, but I would skip thinking about services at the moment.

FWinkler79 commented:
Hi @chombium 

thanks for the clarifications!

> We are proposing cloudfoundry.system.* attribute group for platform components and cloudfoundry.app.* for
CF applications.

I think that is fine. I just understood @oppegard in a way that there can be multiple CF system components (each described with the `cloudfoundry.system.*` attributes) that can form a bigger CF system functionality like "gorouter" etc. These bigger CF system functionalities could then be described by the `service.*` attributes. But maybe I misunderstood this.

Likewise there can be multiple CF apps (each described with the `cloudfoundry.app.*` attributes) that form a bigger (micro-)service. Again, these bigger (micro-) services could then be described using the `service.*` attributes.

> I want to avoid the confusion as in CF there is a concept of [services](https://docs.cloudfoundry.org/services/overview.html), which are basically some resources running outside CF which can be used by a CF application

Services in CF are not that special, imho, and they are closely related to CF apps: 
To implement / provide a service, you need to implement one or more CF apps, integrate them, and expose the (service) functionality they provide with a Service Broker. You register the broker in CF, so that other CF apps can create instances of your service and consume it. Creating a service instance is much like creating a tenant, i.e. it usually reserves resources for a particular consumer in the CF apps implementing the service .
The CF apps implementing the service functionality should know the service they belong to. So each CF app can individually be described using `cloudfoundry.apps.*` attributes, and the service they help provide would be described by the `service.*` attributes. The service instances could then be described via the `service.instance.*` attributes.
Of course, the `service.*` and `service.instance.*` attributes would have to be set by the CF apps in this case - there is no way the CF platform would know which service the CF apps implement.

> I understand a service as it is described in Otel as a group of applications deployed somewhere running together to fulfill some goal. 

Exactly. I think that can be nicely mapped to the "CF apps providing a service" example I described above.

FWinkler79 commented:
@KarstenSchnitter I think you still wanted to add something here, right?

KarstenSchnitter commented:
I think, I have addressed all open questions, but need some feedback from @lmolkova before marking them as resolved. If the current state is acceptable, there is nothing to be changed from my side.

KarstenSchnitter commented:
@joaopgrassi and @lmolkova is there a way we can make progress here? The PR is approved by subject matter experts from the Cloud Foundry community.

joaopgrassi commented:
I will try and take a look at it again in the upcoming days.

joaopgrassi commented on model/registry/cloudfoundry.yaml:
```diff
@@ -0,0 +1,158 @@
+groups:
+  - id: registry.cloudfoundry
+    prefix: cloudfoundry
```
Prefix is not allowed anymore. Please add the value here in all attributes. Like `cloudfoundry.system.id`. This should fix the policy errors

> Group 'cloudfoundry.process' uses prefix 'cloudfoundry.process'. All attribute should be fully qualified with their id, prefix is no longer supported.

KarstenSchnitter replied:
Thanks, will do.

joaopgrassi reviewed (commented):
Please take a look at the checks failing. Also update the issue templates by running `make generate-gh-issue-templates` and `make fix` to update any left over tables.


One general ask I have to the group is: Resource attributes must be immutable - that means a resource is created during the app start up with say CloudFoundry attributes. Once the app is up and running, these attributes cannot change. I wanted to make this clear so you as the experts do the exercise in seeing if there's any such attributes introduced here that are affected by this constraint. 

Another fact is that there's effort currently in a new group for Resource and Entities. The reason I'm bringing it up is since this PR is adding resource attributes I'm not sure if we want to block it. Maybe we allow it to be part of the registry as a first step. 

@jsuereth could you please share your thoughts here?

KarstenSchnitter reviewed (commented):
I took care of the prefixes and hopefully all validation errors. I can confirm, that all resource attributes are immutable. Once an application instance is started, all its processes will have uniquely and immutably defined resource attributes of the proposed set.

KarstenSchnitter commented:
There is an issue on the main branch with the Github link to the registry, which has a 404 error:

https://github.com/open-telemetry/semantic-conventions/blob/b855a36cf4b8a3c5461ad16fa4ace0c615d11c2a/CONTRIBUTING.md?plain=1#L225C15-L225C94

This fails the tests, but is not caused by this PR.

KarstenSchnitter commented:
The latest merge should fix the issue with the main branch.

lmolkova commented on model/cloudfoundry/registry.yaml:
```diff
@@ -0,0 +1,157 @@
+        brief: >
+          A guid or another name describing the event source.
+        note: |
+          CloudFoundry defines the `source_id` in the [Loggegator v2 envelope](https://github.com/cloudfoundry/loggregator-api#v2-envelope).
```
is it a typo? 

```suggestion
          CloudFoundry defines the `source_id` in the [Loggregator v2 envelope](https://github.com/cloudfoundry/loggregator-api#v2-envelope).
```

lmolkova commented on model/cloudfoundry/registry.yaml:
```diff
@@ -0,0 +1,157 @@
+        brief: >
+          A guid describing the concrete instance of the event source.
+        note: |
+          CloudFoundry defines the `instance_id` in the [Loggegator v2 envelope](https://github.com/cloudfoundry/loggregator-api#v2-envelope).
```
a typo?

```suggestion
          CloudFoundry defines the `instance_id` in the [Loggregator v2 envelope](https://github.com/cloudfoundry/loggregator-api#v2-envelope).
```

lmolkova commented on model/cloudfoundry/registry.yaml:
```diff
@@ -0,0 +1,157 @@
+          reported by `cf app <app-name> --guid`.
+        examples: ['218fc5a9-a5f1-4b54-aa05-46717d0ab26d']
+      - id: cloudfoundry.app.instance.id
+        type: string
```
is it intentionally a `string`? Examples and description suggest it could be an `int`. If you want to have string for extensibility/edge cases, it's totally fine too.

KarstenSchnitter replied:
Thanks for that finding. Since we separated the `application.instance.id` from the `system.instance.id`, this could indeed be an `int`. Still I would prefer a `string` exactly for extensibility and edge cases.

lmolkova commented on model/cloudfoundry/resources.yaml:
```diff
@@ -0,0 +1,51 @@
+    attributes:
+      - ref: cloudfoundry.source.type
+        requirement_level:
+          conditionally_required: if the the event is a log message
```
from pure otel side, `event` and `log` are meaningful terms (events are a subset of logs), so this looks confusing. Maybe this would work (if it's correct):

```suggestion
          conditionally_required: if the the source produces log messages
```

Since resource attributes are applied globally to all telemetry reported by a trace/metric/log provider, individual telemetry items (logs) can't have different resource attributes. If source type is a property of a specific log, you may want to define it in a separate `attribute_group` or describe the whole event semantic convention including this attribute.

KarstenSchnitter replied:
This attribute should only be added to logs, not to metrics or spans. It is constant for the service for all log messages. I understand, that this is not the correct way to model this as a resource attribute. I will remove this attribute from this PR.

@lmolkova: What would be your proposal to include this attribute in log messages? Should this be a separate convention just for this log attribute?

lmolkova replied:
If it's a constant for all logs emitted by the service it can be added as a resource attribute and then populated only on the [logger provider](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/sdk.md#loggerprovider) but not spans or metrics - that's totally fine.
It'd be great to explain it in the note/brief so people know what is it and how to use it.

If you're interested in describing it as a log attribute, please take a look at some log examples, e.g. [exceptions](https://github.com/open-telemetry/semantic-conventions/blob/b780f480e73ffc2d1da117689475b7b9c8af610b/docs/exceptions/exceptions-logs.md#attributes).

KarstenSchnitter replied:
I suggest to leave this attribute out for now. It will most likely be populated only inside the CloudFoundry system components by an OTel Collector. As indicated, this field is hard to fill for application developers trying to instrument their own application. Therefore, I think it is better to leave it out and add it at a later point in time, when it is clearer, how it is added and that it is still required in a world, where CloudFoundry primarily uses OpenTelemetry for its observability signals.

KarstenSchnitter reviewed (commented):
@lmolkova Thanks you for the review. I am going to remove the `source.type` with my next commits.

KarstenSchnitter commented:
I removed the `source.type` with the last commit.

lmolkova commented on docs/resource/cloudfoundry.md:
```diff
@@ -0,0 +1,263 @@
+
+## Cloud Foundry System Component
+
+**type:** `cloudfoundry`
+
+**Description:** A CloudFoundry system component, i.e. parts of the runtime.
```
nit: it's duplicated below (it's now auto-generated  from yaml)

```suggestion
```

lmolkova commented on docs/resource/cloudfoundry.md:
```diff
@@ -0,0 +1,263 @@
+
+## Process
+
+**type:** `cloudfoundry.process`
+
+**Description:** A CloudFoundry Application Process
```
nit: it's duplicated below (auto-generated)

```suggestion
```

lmolkova commented on docs/resource/cloudfoundry.md:
```diff
@@ -0,0 +1,263 @@
+
+## Application
+
+***type:** `cloudfoundry.app`
+
+**Description:** A CloudFoundry Application
+
```
nit: auto-generated below 
```suggestion
```

lmolkova commented on docs/resource/cloudfoundry.md:
```diff
@@ -0,0 +1,263 @@
+
+## Space
+
+**type:** `cloudfoundry.space`
+
+**Description:** A CloudFoundry Space
+
```
```suggestion
```

lmolkova commented on docs/resource/cloudfoundry.md:
```diff
@@ -0,0 +1,263 @@
+
+## Organization
+
+**type:** `cloudfoundry.org`
+
+**Description:** A CloudFoundry Organization
+
```
```suggestion
```

lmolkova commented on docs/resource/cloudfoundry.md:
```diff
@@ -0,0 +1,263 @@
+<!-- END AUTOGENERATED TEXT -->
+<!-- endsemconv -->
+
+## Source Type
```
this section should probably be removed since the yaml definition was removed. Not sure why CI checks didn't detect it

KarstenSchnitter replied:
I will take care of that.

lmolkova commented on docs/resource/cloudfoundry.md:
```diff
@@ -0,0 +1,263 @@
+brackets.
+<!-- endsemconv -->
+
+[DocumentStatus]: https://github.com/open-telemetry/opentelemetry-specification/tree/v1.26.0/specification/document-status.md
```
super-nit

```suggestion
[DocumentStatus]: [DocumentStatus]: https://opentelemetry.io/docs/specs/otel/document-status
```

lmolkova reviewed (approved):
LGTM, suggesting a few cosmetic changes

KarstenSchnitter commented:
Thanks everyone for all the help, work, and dedication you provided to include this change into the repository.
